src/mapping.ts

```diff
@@ -44,6 +44,43 @@ export class GenericParseTreeNode extends GenericNode {
 
 registerNodeFactory(ParserRuleContext, () => new GenericParseTreeNode());
 
+/**
+ * Implements a transformation from an ANTLR parse tree (the output of the parser) to an AST (a higher-level
+ * representation of the source code).
+ */
+export class ParseTreeToASTTransformer extends ASTTransformer {
+
+    /**
+     * Performs the transformation of a node and, recursively, its descendants. In addition to the overridden method,
+     * it also assigns the parseTreeNode to the AST node so that it can keep track of its position.
+     * However, a node factory can override the parseTreeNode of the nodes it creates (but not the parent).
+     */
+    transform(source?: any, parent?: Node): Node | undefined {
+        const node = super.transform(source, parent);
+        if (node != undefined && node.origin == undefined && source instanceof ParserRuleContext) {
```

Comment:
Maybe I would relax the test to `node && !node.origin && source instanceof ParserRuleContext`